core/src/main/java/io/grpc/ManagedChannelBuilder.java

```diff
@@ -42,7 +40,8 @@
  * @param <T> The concrete type of this builder.
  */
 public abstract class ManagedChannelBuilder<T extends ManagedChannelBuilder<T>> {
-  public static ManagedChannelBuilder forAddress(String name, int port) {
+  public static <T extends ManagedChannelBuilder<T>> ManagedChannelBuilder<T> forAddress(
```

Comment:
With this PR as it is, I get warnings when compiling.

```
/home/ejona/clients/grpc-java/core/src/test/java/io/grpc/ManagedChannelProviderTest.java:113: warning: [unchecked] builderForAddress(String,int) in BaseProvider overrides <T>builderForAddress(String,int) in ManagedChannelProvider
    protected ManagedChannelBuilder builderForAddress(String host, int port) {
                                    ^
  return type requires unchecked conversion from ManagedChannelBuilder to ManagedChannelBuilder<T>
  where T is a type-variable:
    T extends ManagedChannelBuilder<T> declared in method <T>builderForAddress(String,int)
```

I do think the generics aren't right in this PR, in that the following would compile without any warnings:
```
ManagedChannelBuilder<NettyChannelBuilder> mcb
    = ManagedChannelBuilder.forAddress("localhost", 443);
```

The generics seem to behave nicely with the `<?>`; they don't needlessly throw away information. I think it is important to consider what the user would store the type as. This seems reasonable:

```java
ManagedChannelBuilder<?> mcb = ManagedChannelBuilder.forAddress("localhost", 443);
```

As proof that that is usable, the below compiles with no warnings.

```java
public class Blah {
  public static void main(String[] args) {
    A<?> a = new B();
    A<?> a2 = a.get();
    A<?> a3 = a2.get();
  }

  private static abstract class A<T extends A> {
    public abstract T get();
  }

  private static class B extends A<B> {
    public B get() {
      return this;
    }
  }
}
```